File: NeetCode 150/Heap/355. Design Twitter.py
class Twitter:

    def __init__(self):
        self.time = 0
        self.tweets = defaultdict(list) # userId -> list of (time, tweetId)
        self.following = defaultdict(set) # userId -> set of followees
        

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweets[userId].append((self.time,tweetId))
        self.time -= 1 # decreasing time ensures newer tweets come first

    def getNewsFeed(self, userId: int) -> List[int]:
        res =[]
        heap = []

         # user should follow himself
        self.following[userId].add(userId)
        # push latest tweet of each followed user
        for followee in self.following[userId]:
            if self.tweets[followee]:
                time , tweetID = self.tweets[followee][-1]
                idx = len(self.tweets[followee]) - 1
                heapq.heappush(heap,(time, tweetID, followee, idx))   

        # get 10 most recent tweets
        while heap and len(res) < 10:
            time, tweetid, user, idx = heapq.heappop(heap)
            res.append(tweetid) 

            # move to previous tweet of same user
            if idx > 0:
                next_time,next_tweetid = self.tweets[user][idx -1]
                heapq.heappush(heap,(next_time,next_tweetid,user,idx-1))
        return res

    def follow(self, followerId: int, followeeId: int) -> None:
        self.following[followerId].add(followeeId)
        

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followerId != followeeId:
            self.following[followerId].discard(followeeId)
            #does not throw error if key not present else remove throws error
        


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)

File: NeedCode 150 (JAVA)/Heaps/355. Design Twitter.java
class Twitter {
    //why static : The keyword static means Tweet does not need a reference to the enclosing Twitter object to save memory.
    /*
    Conceptually, Java treats it like:

class Tweet {
    Twitter this$0;   // hidden reference
    int id;
    int time;
}
it doesn't need twitter reference here
    */
    private static class Tweet{
        int id;
        int time;

        Tweet(int id, int time){
            this.id = id;
            this.time = time;
        }
    }

    private int time;
    private Map<Integer, Set<Integer>> followMap; //User has a set of followee(whom he follows)
    private Map<Integer, List<Tweet>> tweetMap; //for every id, list of Tweet class

    public Twitter() {
        time = 0;
        followMap = new HashMap<>();
        tweetMap = new HashMap<>();
    }
    
    public void postTweet(int userId, int tweetId) {
        tweetMap.putIfAbsent(userId, new ArrayList<>());
        tweetMap.get(userId).add(new Tweet(tweetId,time++));
    }
    
    public List<Integer> getNewsFeed(int userId) {
        PriorityQueue<Tweet> maxHeap = new PriorityQueue<>((a,b) -> b.time - a.time);

        //own Tweets
        if(tweetMap.containsKey(userId)){
            for(Tweet t : tweetMap.get(userId)){
                maxHeap.offer(t);
            }
        }

        //tweets from followees (whom he follows)
        if(followMap.containsKey(userId)){
            for(int followee: followMap.get(userId)){
                if(tweetMap.containsKey(followee)){
                    for(Tweet t: tweetMap.get(followee)){
                        maxHeap.offer(t);
                    }
                }
            }
        }

        List<Integer> feed = new ArrayList<>();
        while(!maxHeap.isEmpty() && feed.size() < 10){
            feed.add(maxHeap.poll().id);
        }
        
        return feed;
    }
    
    public void follow(int followerId, int followeeId) {
        if (followerId == followeeId) return;

        followMap.putIfAbsent(followerId, new HashSet<>());
        followMap.get(followerId).add(followeeId);
    }
    
    public void unfollow(int followerId, int followeeId) {
        if(followMap.containsKey(followerId))
            followMap.get(followerId).remove(followeeId);
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */
